### PR TITLE
fix(server): `unstable_httpBatchStreamLink` hiccups

### PR DESCRIPTION
**Pull request by KATT**
Closes #

## 🎯 Changes

Introduced bugs in some edge cases in #6206 that were show-cased when reducing timeout for iterations in the tests

## Summary

- Reintroduce deferreds in jsonl consumer
- Add `using` a bit more in tests, for fun

## Notes

Hide whitespace when reviewing

CleanShot 2024-11-15 at 12 54 12@2x

## ✅ Checklist

- [ ] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/server/src/unstable-core-do-not-import/stream/jsonl.ts
Comment (KATT) [inline packages/server/src/unstable-core-do-not-import/stream/jsonl.ts:1]: cc @oleasteo 

I think we could refactor this whole file to leverage async iterables for the production of the stream.... and some in the consumer now

i've made a bit of spaghetti right now 😳

File: packages/server/src/unstable-core-do-not-import/stream/sse.ts [hunk 1]
```diff
@@ -32,7 +32,7 @@ export interface PingOptions {
 export interface SSEStreamProducerOptions<TValue = unknown> {
   serialize?: Serialize;
   data: AsyncIterable<TValue>;
-  abortCtrl: AbortController;
```
Comment (KATT) [inline packages/server/src/unstable-core-do-not-import/stream/sse.ts:35]: This shouldn't be needed anymore

We have this: https://github.com/trpc/trpc/blob/8b37fed52bc53acea6df9bb0896a8cdc89f8e39c/packages/server/src/adapters/node-http/writeResponse.ts#L47-L49

File: packages/server/src/unstable-core-do-not-import/stream/utils/asyncIterable.ts [hunk 2]
```diff
@@ -81,8 +77,6 @@ export async function* takeWithGrace<T>(
       yield result.value;
       if (--count === 0) {
         timerPromise = timer.start();
-        // eslint-disable-next-line @typescript-eslint/no-floating-promises
-        timerPromise.then(() => opts.abortCtrl.abort());
```
Comment (KATT) [inline packages/server/src/unstable-core-do-not-import/stream/utils/asyncIterable.ts:85]: Shouldn't be needed as ending the stream will cause the abort controller to be triggered
